Improve ANR error message information

fractalwrench opened the pull request:
## Goal

Android records a cause for an ANR in the `ProcessRecord` for the app process. Adding the message to error reports will improve context and differentiate timed out input dispatching from failing to respond to a broadcast event.

## Changeset

When `SIGQUIT` is caught bugsnag creates an `Error` which captures the stacktrace. Rather than sending this report immediately, the notifier now polls the `ActivityManager` for 30 seconds in an attempt to access [process error information](https://android.googlesource.com/platform/frameworks/base/+/refs/heads/master/services/core/java/com/android/server/am/AppErrors.java#1088).

Once this information is available the notifier updates the error message to the value of `ProcessErrorStateInfo#shortMsg`, resulting in a stacktrace that looks like this:

```
ANR:  Input dispatching timed out (Waiting to send non-key event because the touched window has not finished processing certain input events that were delivered to it over 500.0ms ago.  Wait queue length: 24.  Wait queue head age: 6396.4ms.)
        at java.lang.Thread.sleep(Thread.java:-2)
        at java.lang.Thread.sleep(Thread.java:373)
        at java.lang.Thread.sleep(Thread.java:314)
        at com.bugsnag.android.example.ExampleActivity$onCreate$3.onClick(ExampleActivity.kt:38)
        at android.view.View.performClick(View.java:6256)
```

*Important Note*: Android has a concept of 'silent ANRs' which occur when the process is in the background. For silent ANRs, the process is [killed immediately](https://android.googlesource.com/platform/frameworks/base/+/refs/heads/master/services/core/java/com/android/server/am/AppErrors.java#1082) without showing a dialog to the user. With this changeset Bugsnag will no longer detect silent ANRs, which may reduce 'overreporting' that has been raised as feedback.

### Other changes

- Made it possible to set the `name` and `message` on a `BugsnagException`, and added public methods for doing this via the `Error` object also
- Updated `Error` and `Exceptions` to always use a `BugsnagException` field. Regular `Throwables` are now wrapped into a `BugsnagException`

## Tests

- Added unit test coverage
- Ran existing tests and manually tested ANR detection in example app

kattrali commented on bugsnag-plugin-android-anr/src/main/java/com/bugsnag/android/AnrPlugin.kt:
```diff
@@ -16,12 +22,45 @@ internal class AnrPlugin : BugsnagPlugin {
+            .severityReasonType(HandledState.REASON_ANR)
+            .build()
+
+        // wait for the ANR dialog to show, reporting once the information is available
```
The [info is available](https://android.googlesource.com/platform/frameworks/base/+/refs/heads/master/services/core/java/com/android/server/am/AppErrors.java#1088) before [the dialog shows](https://android.googlesource.com/platform/frameworks/base/+/refs/heads/master/services/core/java/com/android/server/am/AppErrors.java#1093), but just barely. The delay is caused by dumping various stats about the running processes.

kattrali commented on bugsnag-plugin-android-anr/src/main/java/com/bugsnag/android/AnrPlugin.kt:
```diff
@@ -16,12 +22,45 @@ internal class AnrPlugin : BugsnagPlugin {
+
+                        // TODO determine how to populate error report
+                        val metaData = error.metaData
+                        metaData.addToTab("ANR", "Message", anrDetails.shortMsg)
```
This should replace the `errMsg` above.

kattrali commented on bugsnag-plugin-android-anr/src/main/java/com/bugsnag/android/AnrPlugin.kt:
```diff
@@ -16,12 +22,45 @@ internal class AnrPlugin : BugsnagPlugin {
+                        // TODO determine how to populate error report
+                        val metaData = error.metaData
+                        metaData.addToTab("ANR", "Message", anrDetails.shortMsg)
+                        metaData.addToTab("ANR", "Component", anrDetails.tag)
```
Isn't this the same as the default context, since it shows the activity name? If not, is it a better context for what is there currently?

fractalwrench replied:
I don't think this is the same as the default context. I believe if an ANR occurs in a `BroadcastReceiver` this value will be different - although we'll need to test that hypothesis.

kattrali replied:
That would be pretty cool if so!

fractalwrench replied:
This does not appear to be the case in my testing with a broadcastReceiver.

kattrali commented on bugsnag-plugin-android-anr/src/main/java/com/bugsnag/android/AnrPlugin.kt:
```diff
@@ -4,8 +4,14 @@ import java.nio.ByteBuffer
 internal class AnrPlugin : BugsnagPlugin {
 
+    companion object {
+        private const val INFO_POLL_THRESHOLD_MS: Long = 2500
```
This might need to be a shorter threshold. I've seen the dialog pop in ~1 second to as many as 12 seconds (when the device was overheating from other processes).

kattrali commented on bugsnag-plugin-android-anr/src/androidTest/java/com/bugsnag/android/AnrDetailsCollectorTest.kt:
```diff
@@ -0,0 +1,75 @@
+    fun differentPidReturnsNull() {
+        Mockito.`when`(am.processesInErrorState).thenReturn(listOf(stateInfo))
+        val captureProcessErrorState = collector.captureProcessErrorState(am, PID_EXAMPLE)
+        assertEquals(stateInfo, captureProcessErrorState)
```
This test and the one below seem to be identical. Was that intentional? Nothing here is returning null.

fractalwrench replied:
This was unintentional - I've updated the assertion for this test

kattrali commented on bugsnag-plugin-android-anr/src/main/java/com/bugsnag/android/AnrDetailsCollector.kt:
```diff
@@ -0,0 +1,73 @@
+        }
+    }
+
+    internal fun mutateError(error: Error, anrState: ProcessErrorStateInfo) {
```
This function could use a more descriptive name about how the error is changed. `addErrorStateInfo()`, etc.

kattrali commented on bugsnag-plugin-android-anr/src/main/java/com/bugsnag/android/AnrDetailsCollector.kt:
```diff
@@ -0,0 +1,73 @@
+internal class AnrDetailsCollector {
+
+    companion object {
+        private const val INFO_POLL_THRESHOLD_MS: Long = 1000
```
Did you try this at a shorter duration? 100ms should be acceptable for the short period of time this poll needs to run.

fractalwrench commented:
@kattrali thanks for the review comments, I've updated the implementation and also added an `ANR` tab with a `CPU Info` key. The poll interval in `AppNotRespondingMonitor` has been left at 5ms.

tomlongridge commented on bugsnag-android-core/src/androidTest/java/com/bugsnag/android/BugsnagExceptionTest.kt:
```diff
@@ -0,0 +1,82 @@
+        return exc
+    }
+
+    // TODO convert JavaScriptException to BugsnagException and create Stacktrace
```
Any reason this can't be done now? Is it lengthy to do?

fractalwrench replied:
I added this comment early on in the implementation of this fix and evidently forgot to remove it. It turns out that `JavaScriptException` adds in an additional `columnNumber` field in its serialisation code, so it can't use `BugsnagException` unless we update the implementation of that class to handle this. I also felt that converting `JavaScriptException` to `BugsnagException` would make this changeset larger than it needed to be to fix the issue, and risk introducing some other issue.
